[MERGE] *: add Quick Edit and Auto Save

mcm-odoo opened the pull request:
This PR adds two main features: quick edit and auto save

Quick Edit
---

The quick edit allows to click on fields in readonly form view to switch
into edit mode. After switching mode, the clicked field is automatically
focused.
A few fields have a custom quick edit behaviour after mode switched:
  - checkboxes automatically toggle.
  - radio buttons are set to the selected value.
  - one2many list's cell are focused.

One2many list fields now show the "add a line" in readonly mode.

In order to minify the differences between edit and readonly modes
multiple things have been done done:
- a lot of `o_edit_only`,`o_read_only` labels have been removed and
  `placeholder` have been set on fields in a lot of xml templates.
- the render function of many2one and x2many radio
   fields have been changed to render them the same in edit mode and read mode.
- margins in inner form groups have been removed.
- a minimum height on rows has been set to align them.
- empty fields are now visible. (as a blank line)

Auto Save
---

Auto save is self explanatory.
With auto save, changing the pager, going back in the breadcrumb,
going to another action or clicking on a menu item doesn't ask to
confirm changes if any but automatically saves them.

In settings, the confirm dialog has been revamped.
We can now decide to "Save" or "Discard" the changes or "Stay Here" to
do nothing.

---

[Enterprise PR](https://github.com/odoo/enterprise/pull/14931)
task 2330101

aab-odoo commented on addons/web/static/src/js/views/form/form_renderer.js:
```diff
@@ -1112,6 +1138,31 @@ var FormRenderer = BasicRenderer.extend({
+        if (this.mode === 'edit') {
+            return;
+        }
+        if (ev.currentTarget.closest('.oe_subtotal_footer')) {
```
weird

aab-odoo replied:
still weird :-)

aab-odoo commented on addons/web/static/tests/fields/relational_fields/field_many2one_tests.js:
```diff
@@ -393,7 +393,7 @@ QUnit.module('fields', {}, function () {
                         '<field name="turtle_trululu" context="{\'show_address\': 1}" options="{\'always_reload\': True}"/>' +
                         '</form>',
-                    "turtle,false,list": '<tree editable="bottom">' +
+                    "turtle,false,list": '<tree>' +
```
???

aab-odoo replied:
Why do you need to change this?

aab-odoo commented on addons/web/static/src/js/fields/relational_fields.js:
```diff
@@ -515,6 +515,11 @@ var FieldMany2One = AbstractField.extend({
+        const escapedValue = _.escape(`${value}`.trim());
+        const [, ...lines] = escapedValue.split('\n');
+        const extraText = lines.map((line) => `<span>${line}</span>`).join('<br/>');
+        this.$('.o_field_many2one_extra').html(extraText);
```
same as in readonly? factorize the code then. Note that building html in JS isn't what we like to see (xss risks)

aab-odoo commented on addons/web/static/src/js/fields/relational_fields.js:
```diff
@@ -1921,6 +1940,46 @@ var FieldOne2Many = FieldX2Many.extend({
+                }).guardedCatch(function() {
+                    self.creatingRecord = false;
+                })
+                ;
```
;

aab-odoo commented on addons/sale_stock/static/src/js/qty_at_date_widget.js:
```diff
@@ -125,7 +125,13 @@ var QtyAtDateWidget = Widget.extend({
+     * @param {MouseEvent} ev
+     */
+    _onClickButton: function (ev) {
+        ev.stopPropagation();
```
red flag?

aab-odoo commented on addons/web_editor/static/src/js/backend/field_html.js:
```diff
@@ -296,6 +296,9 @@ var FieldHtml = basic_fields.DebouncedField.extend(TranslatableFieldMixin, {
             container.innerHTML = value;
+            $(container).find('[href]').on('click', ev => {
+                ev.preventDefault();
+            });
```
red flag

aab-odoo reviewed (commented):
This PR scares me. I'm not saying it is bad, but the diff is huge and could introduce a shitload of tricky issues. Also, I'm not sure about the autosave part, spec-wise.

aab-odoo reviewed (commented):
Commit message quick edit:
"After switching mode, **the clicked in automatically**
focused"
And the rest of the message should be adapted as you split the commit

aab-odoo commented on addons/web/static/src/js/fields/relational_fields.js:
```diff
@@ -1157,6 +1178,21 @@ var FieldX2Many = AbstractField.extend(WidgetAdapterMixin, {
+    doQuickEdit: function (extraInfo) {
+        this._super(...arguments);
+
+        if (extraInfo.subFieldName) {
```
I'm a bit concerned. In FieldBoolean, you check the mode in this method, but in Many2one and X2many you don't. I think it will crash if we get here in readonly (which I assume happens if the field is `readonly="1"`). If I'm right, this shows as I said above that the check should be in AbstractField

aab-odoo commented on addons/web/static/src/js/fields/relational_fields.js:
```diff
@@ -2579,6 +2641,35 @@ var FormFieldMany2ManyTags = FieldMany2ManyTags.extend({
+     */
+    doQuickEdit: function () {
+        this._super(...arguments);
+        this.many2one.$input.click();
```
isn't this what `activate` does?

aab-odoo commented on addons/web/static/src/js/fields/relational_fields.js:
```diff
@@ -2579,6 +2641,35 @@ var FormFieldMany2ManyTags = FieldMany2ManyTags.extend({
+     */
+    _triggerQuickEdit: function (ev) {
+        if (!this.nodeOptions.no_edit_color &&
+            !ev.target.closest('.dropdown-toggle')
```
this method should only be called in readonly

aab-odoo commented on addons/web/static/src/js/views/basic/basic_renderer.js:
```diff
@@ -727,6 +727,10 @@ var BasicRenderer = AbstractRenderer.extend(WidgetAdapterMixin, {
+            // Distinct readonly from renderer and readonly from modifier,
+            // renderer can be readonly while modifier not.
+            // This is needed as modifiers are updated after first render
+            readonlyModifier: modifiers.readonly,
```
do you update this option during the lifecycle of the widget?

aab-odoo commented on addons/web/static/src/js/views/form/form_controller.js:
```diff
@@ -34,6 +35,7 @@ var FormController = BasicController.extend({
         this.defaultButtons = params.defaultButtons;
         this.hasActionMenus = params.hasActionMenus;
         this.toolbarActions = params.toolbarActions || {};
+        this._autoSave = false;
```
should not be useful anymore (w.r.t. the last specs)

aab-odoo commented on addons/web/static/src/js/views/form/form_renderer.js:
```diff
@@ -1086,7 +1105,22 @@ var FormRenderer = BasicRenderer.extend({
+                !excludedAncestorQueries.some(x => widget.el.closest(x))
+            ) {
+                widget.setQuickEdit();
+            }
```
don't understand this part

aab-odoo commented on addons/web/static/src/js/views/form/form_renderer.js:
```diff
@@ -1149,6 +1208,9 @@ var FormRenderer = BasicRenderer.extend({
     _onNotebookTabChanged: function () {
         core.bus.trigger('DOM_updated');
     },
+    _onQuickEdit: function (ev) {
```
missing docstring (here + several other places)

aab-odoo commented on addons/web/static/src/js/views/list/list_renderer.js:
```diff
@@ -568,7 +568,7 @@ var ListRenderer = BasicRenderer.extend({
             title = formatter(value, field, _.extend(formatOptions, {escape: false}));
         }
-        return $td.html(formattedValue).attr('title', title);
+        return $td.html(formattedValue).attr('title', title).attr('name', name);
```
?

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+
+        await testUtils.dom.click(form.$('.o_field_widget[name="display_name"]'));
+
+        assert.containsOnce(form, '.o_form_view.o_form_editable');
```
could you check the activeElement as well?

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+
+        await testUtils.dom.click(form.$('.o_field_widget[name="priority"] a:first'));
+
+        assert.containsOnce(form, '.o_form_view.o_form_readonly');
```
check the priority field has properly reacted to the click

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+
+        await testUtils.dom.click(form.$('.o_form_label:first'));
+
+
```
extra line

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+        assert.containsOnce(form, '.o_field_boolean input:checked');
+
+        await testUtils.dom.click(form.$('.o_field_widget[name="bar"]'));
+        await testUtils.nextTick();
```
use `owlCompatibilityExtraNextTick` in these cases

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+            res_id: 1,
+        });
+
+        assert.containsOnce(form, '.o_form_view.o_form_readonly');
```
check that Add a line is not there

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+        assert.containsOnce(form, '.o_form_view.o_form_readonly');
+
+        await testUtils.dom.click(form.$('.o_field_widget[name="date"]'));
+        await testUtils.nextTick();
```
owlCompatibilityExtraNextTick

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
         assert.verifySteps(['mounted', 'willUnmount']);
     });
+
+    QUnit.module('Form View: Quick Edition');
```
instead of a sub module, I would prefix all tests with "Quick Edition:". That way, you can still filter in the tests page to only execute those tests. Having submodules in tests files often leads to a mess.

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+            model: 'partner',
+            data: this.data,
+            arch: `
+                <form string="Partners">
```
you don't need to change that, but so you know: `string="Partners"` is useless

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+            res_id: 1,
+        });
+
+        assert.containsOnce(form, '.o_form_view.o_form_readonly');
```
check the selected value

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+        form.destroy();
+    });
+
+    QUnit.test('Auto save on page change', async function (assert) {
```
should not be prefixed by "Quick edit:" as this is another feature

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+        assert.containsOnce(form, '.o_form_view.o_form_readonly');
+
+        await testUtils.dom.click(form.$('.o_field_widget[name="display_name"]'));
+        await testUtils.nextTick();
```
is this necessary?

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+
+        await testUtils.controlPanel.pagerPrevious(form);
+        assert.containsOnce(form, '.o_form_view.o_form_editable');
+        assert.strictEqual(form.$('input.o_field_widget[name="display_name"]').val(), 'test');
```
maybe add an assert.step in mockRPC to log calls to "save"

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9665,6 +9665,399 @@ QUnit.module('Views', {
+        assert.strictEqual(actionManager.$('.o_field_widget[name="display_name"]').text().trim(), 'first record');
+
+        await testUtils.dom.click(actionManager.$('.o_field_widget[name="display_name"]'));
+        await testUtils.nextTick();
```
why?

aab-odoo commented on addons/web/static/src/js/views/basic/basic_model.js:
```diff
@@ -1618,16 +1618,22 @@ var BasicModel = AbstractModel.extend({
         }
         var rel_data = _.pick(data, 'id', 'display_name');
 
+        const context = this._getContext(record, {
```
looks like you only use this if the `if ()`

aab-odoo commented on addons/web/static/src/scss/fields.scss:
```diff
@@ -87,6 +87,21 @@
+            padding-top: 0;
+            padding-bottom: 0;
+        }
+    }
```
Your changes only apply to form views, right? Here you define rules for everywhere

aab-odoo commented on addons/web/static/tests/fields/relational_fields_tests.js:
```diff
@@ -2165,8 +2165,9 @@ QUnit.module('relational_fields', {
-            "value should be 'Black'");
+        assert.strictEqual(form.$('.o_field_widget').text().trim().split(/\s+/g).join(','), 'Red,Black');
+        assert.containsOnce(form, '.o_radio_input[data-index=1]:checked',
+            "'Black' should be checked");
```
the diff in tests looks a bit weak with respect to the diff in the code. Some changes are probably worth new tests (like the many2one multi line stuff, and the Add a line in x2many)

aab-odoo commented on addons/web/static/src/js/views/basic/basic_controller.js:
```diff
@@ -80,28 +80,69 @@ var BasicController = AbstractController.extend(FieldManagerMixin, {
      *          rejected otherwise
      */
-    canBeDiscarded: function (recordID) {
+    canBeDiscarded: function (recordID, options={}) {
```
I thought we only wanted this for the Settings form view? @odoongr

aab-odoo reviewed (commented):
Commit message "[IMP] dicard changes confirmation dialog improvement" needs to be properly formatted. Also, I didn't review that commit yet as i thought we wanted this only in the settings form view.

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -3123,22 +3113,18 @@ QUnit.module('Views', {
+        assert.strictEqual(form.$('input[name=foo]').val(), 'new value', "input should contain new value");
 
-        assert.strictEqual(nbWrite, 0, 'no write RPC should have been done');
+        assert.strictEqual(nbWrite, 1, 'no write RPC should have been done');
```
be careful with messages

aab-odoo reviewed (commented):
Up to autosave

aab-odoo commented on addons/web/static/src/js/views/form/form_renderer.js:
```diff
@@ -74,6 +83,16 @@ var FormRenderer = BasicRenderer.extend({
+            const { fieldName, extraInfo } = this.quickEditInfo;
+            this.quickEditInfo = null;
+            const fieldWidget = this.allFieldWidgets[this.state.id]
+                .find(field => field[symbol] === fieldName);
```
as far as i understand, `field[symbol]` is the `node.id`. How can this match with a `fieldName`?

aab-odoo replied:
`node.id` is the fieldName for the first occurrence of a field in the view, by for other occurrences it is the fieldName with a unique suffix...

aab-odoo commented on addons/web/static/src/js/views/form/form_renderer.js:
```diff
@@ -1149,6 +1192,19 @@ var FormRenderer = BasicRenderer.extend({
+        } else {
+            ev.stopPropagation();
+        }
+    },
```
let me summarize the flow as I understand it:
 1) a field is clicked, it triggers 'quick_edit' with some info
 2) the event goes through the renderer which conditionally stops it, and if not, stores that info
 3) the event is catched by the controller which switches to edit mode, re-renders the view, and calls autofocus
 4) in autofocus, we detect that we come from a quick edit flow, and in this case we find the field and call executeQuickEdit

Wouldn't it be simpler to:
 1) a field is clicked, it triggers 'quick_edit' with some info
 2) the event goes up to the controller directly, which conditionally decides to ignore it (with the exclusion selectors), and if not, switches to edit more and **then** calls a specific method of the renderer with the event info
 3) in the renderer, that method executes the quick edit on the correct field

Note that I'm not sure to understand why we need that exclusion selector list in the renderer/controller

aab-odoo commented on addons/web/static/tests/chrome/action_manager_tests.js:
```diff
@@ -550,6 +550,7 @@ QUnit.module('ActionManager', {
         assert.containsOnce(actionManager, '.o_form_view');
 
         await testUtils.dom.click(actionManager.$('.o_form_view .o_data_row:first'));
+        await testUtils.owlCompatibilityExtraNextTick(); // wait for quick edit
```
?

aab-odoo commented on addons/web/static/tests/fields/basic_fields_tests.js:
```diff
@@ -3027,15 +3075,25 @@ QUnit.module('basic_fields', {
         this.data.partner.records[5].timmy = [12];
 
         const form = await createView({
+            debug: 1,
```
:-)

aab-odoo commented on addons/web/static/tests/fields/relational_fields/field_one2many_tests.js:
```diff
@@ -4454,10 +4451,10 @@ QUnit.module('fields', {}, function () {
             });
 
             await testUtils.dom.click(form.$('.o_data_row:first'));
+            await testUtils.owlCompatibilityExtraNextTick(); // wait for quick edit
```
?

aab-odoo commented on addons/web/static/tests/fields/relational_fields/field_one2many_tests.js:
```diff
@@ -5231,6 +5223,7 @@ QUnit.module('fields', {}, function () {
                 'should display the 2 turtles');
 
             await testUtils.dom.click(form.$('.o_data_row:first'));
+            await testUtils.owlCompatibilityExtraNextTick(); // wait for quick edit
```
?

aab-odoo commented on addons/web/static/tests/fields/relational_fields/field_one2many_tests.js:
```diff
@@ -6598,15 +6591,9 @@ QUnit.module('fields', {}, function () {
-            await testUtils.form.clickEdit(form);
-            await testUtils.dom.click(form.$('.o_data_row td:first'));
-
+            await testUtils.owlCompatibilityExtraNextTick();
```
?

aab-odoo commented on addons/web/static/tests/fields/relational_fields/field_one2many_tests.js:
```diff
@@ -8155,6 +8142,7 @@ QUnit.module('fields', {}, function () {
                 'there should 2 and only 2 onchange from closing the partner modal');
 
             await testUtils.dom.click(form.$('.o_data_row:eq(0)'));
+            await testUtils.owlCompatibilityExtraNextTick(); // wait for quick edit
```
?

aab-odoo commented on addons/web/static/tests/fields/relational_fields/field_one2many_tests.js:
```diff
@@ -6598,15 +6591,9 @@ QUnit.module('fields', {}, function () {
-
-            // switch to edit mode and open it again
-            await testUtils.form.clickEdit(form);
-            await testUtils.dom.click(form.$('.o_data_row td:first'));
```
mouarf here you completely change the semantics of the test. Opening the form view first is important. Is there a way to keep opening it?

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9849,6 +9852,557 @@ QUnit.module('Views', {
+
+        await testUtils.dom.click(form.$('.o_form_label:first'));
+
+        assert.containsOnce(form, '.o_form_view.o_form_editable');
```
also check the focus?

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9849,6 +9852,557 @@ QUnit.module('Views', {
+        assert.containsNone(document.body, '.modal');
+
+        await testUtils.dom.click(form.$('.o_field_x2many_list_row_add a'));
+        await testUtils.owlCompatibilityExtraNextTick();
```
do you really need all these extra next tick? owl isn't involved here right??

aab-odoo reviewed (commented):
Good job overall. I left some questions, because I didn't understand everything. We can talk about it.

aab-odoo commented on addons/web/static/src/js/views/basic/basic_controller.js:
```diff
@@ -665,6 +679,22 @@ var BasicController = AbstractController.extend(FieldManagerMixin, {
+     */
+    _onCloseBrowser: function (ev) {
+        if (this.canBeSaved()) {
+            this.saveChanges(this.handle);
```
what if the 'write' rpc crashes?

aab-odoo commented on addons/web/static/tests/views/form_tests.js:
```diff
@@ -9853,6 +9853,65 @@ QUnit.module('Views', {
+
+        await testUtils.fields.editInput(form.$('.o_field_widget[name="display_name"]'), 'test');
+        window.dispatchEvent(new Event("beforeunload"));
+        await testUtils.nextTick();
```
mockRPC to check that it has been saved? Typically, after the beforeunload there is nothing left to do as the tab is supposed to be closed :-)

aab-odoo replied:
also assert that the event isn't prevented

mcm-odoo commented on addons/web/static/src/js/views/basic/basic_controller.js:
```diff
@@ -155,6 +134,26 @@ var BasicController = AbstractController.extend(FieldManagerMixin, {
+                    reload: false,
+                });
+            } catch (e) {
+                return Promise.reject();
```
remove try/catch

mcm-odoo commented on addons/web/static/src/js/views/form/form_controller.js:
```diff
@@ -78,6 +82,21 @@ var FormController = BasicController.extend({
+     */
+    canBeDiscarded: function (recordId) {
+        if (!recordId || recordId === this.handle || !this.isDirty(recordId)) {
+            return Promise.resolve(true);
```
check always `return Promise.resolve(true);`

mcm-odoo commented on addons/web/static/tests/chrome/action_manager_tests.js:
```diff
@@ -3816,30 +3812,30 @@ QUnit.module('ActionManager', {
-        assert.containsOnce(actionManager, '.o_form_view.o_form_editable');
+
+        // should be readonly and so saved
+        assert.containsOnce(actionManager, '.o_form_view.o_form_readonly');
```
add `display_name` in the test

mcm-odoo replied:
and check the breadcrumb

mcm-odoo commented on addons/web/static/src/js/fields/abstract_field.js:
```diff
@@ -508,6 +555,8 @@ var AbstractField = Widget.extend({
      * @param {boolean} [options.forceChange=false] if true, the change event will be
      *   triggered even if the new value is the same as the old one
+     * @param {boolean} [options.forceSave=true] if true, the changes will be saved
+     *   directly. True by default if the view is readonly.
```
try to remove forceSave

mcm-odoo commented on addons/web/static/src/js/fields/abstract_field_owl.js:
```diff
@@ -643,6 +703,18 @@ odoo.define('web.AbstractFieldOwl', function (require) {
+     * List of selector used to not trigger the quick_edit event when clicked
+     * in an element matching one of these selectors.
+     */
+    AbstractField.quickEditionExclusion = [];
```
quickEditExclusion

mcm-odoo commented on addons/web/static/src/js/fields/basic_fields.js:
```diff
@@ -885,6 +888,16 @@ var FieldDate = InputField.extend({
+    _executeQuickEdit: function () {
+        this._super(...arguments);
+        if (this.datewidget) {
+            this.datewidget.$input.click();
```
check: do not call super?

mcm-odoo commented on addons/web/static/src/js/fields/basic_fields.js:
```diff
@@ -3362,6 +3399,7 @@ var FieldDomain = AbstractField.extend({
      */
     _onShowSelectionButtonClick: function (e) {
         e.preventDefault();
+        e.stopPropagation();
```
check: remove it?

mcm-odoo commented on addons/web/static/src/js/fields/relational_fields.js:
```diff
@@ -442,6 +450,18 @@ var FieldMany2One = AbstractField.extend({
+     */
+    _openAutoComplete: function () {
+        if (this.$input.autocomplete("widget").is(":visible")) {
+            this.$input.autocomplete("close");
```
rename function _toggleAutoComplete

aab-odoo commented on addons/web/static/src/js/views/list/list_controller.js:
```diff
@@ -888,6 +888,13 @@ var ListController = BasicController.extend({
+     * @private
+     */
+    _onOdooLeave: function () {
+        this.renderer.unselectRow();
```
You take the usual path here, so you'll wait for onchanges, and thus the record might not be saved. Am I missing something?

To reproduce: have a list view with two fields A and B, and an onchange on A which changes B, set A, close the tab.

aab-odoo commented on addons/web/static/tests/views/list_tests.js:
```diff
@@ -11299,6 +11299,67 @@ QUnit.module('Views', {
+        assert.verifySteps([], 'should not save because of invalid field');
+
+        list.destroy();
+    });
```
there are a lot of tests, which is good, but I think at least two tests are missing: the ones involving onchanges, one for form, one form list. I bet that the latter won't pass.

aab-odoo replied:
also, I'd keep this commit apart.

aab-odoo commented on addons/web/static/src/js/chrome/web_client.js:
```diff
@@ -127,42 +127,40 @@ return AbstractWebClient.extend({
-                                var menu_id = self.menu.action_id_to_primary_menu_id(action.id);
-                                core.bus.trigger('change_menu_section', menu_id);
-                            }
+        var stringstate = $.bbq.getState(false);
```
maybe keep the code indented so reduce the diff here (it's a scary part :-), and we re-writing it anyway)

aab-odoo commented on addons/web/static/src/js/views/form/form_controller.js:
```diff
@@ -677,7 +688,7 @@ var FormController = BasicController.extend({
     _onQuickEdit: function (ev) {
         ev.stopPropagation();
-        this._onEdit();
+        this._setEditMode(ev.data);
```
```js
await this._setEditMode();
this.renderer.quickEdit(quickEditInfo);
```

aab-odoo commented on addons/web/static/src/js/views/form/form_controller.js:
```diff
@@ -437,6 +437,24 @@ var FormController = BasicController.extend({
+        this._disableButtons();
+        // wait for potential pending changes to be saved (done with widgets
+        // allowing to edit in readonly)
+        this.mutex.getUnlockedDef()
```
`return this.mutex.getUnlockedDef()`

and adapt docstring

aab-odoo commented on addons/web/static/src/js/views/form/form_controller.js:
```diff
@@ -437,6 +437,24 @@ var FormController = BasicController.extend({
+        // allowing to edit in readonly)
+        this.mutex.getUnlockedDef()
+            .then(this._setMode.bind(this, 'edit'))
+            .then(() => {
```
remove this then handler (see below)

aab-odoo commented on addons/web/static/src/js/views/form/form_renderer.js:
```diff
@@ -213,8 +213,19 @@ var FormRenderer = BasicRenderer.extend({
+                isSheetBg: true,
+            };
+        } else {
+            const oContent = this.el.closest('.o_content');
```
should not be the concern of the form renderer (hint: looks outside of its DOM)

aab-odoo commented on addons/web/static/src/js/views/form/form_renderer.js:
```diff
@@ -240,13 +240,13 @@ var FormRenderer = BasicRenderer.extend({
+        const { fieldName, extraInfo } = quickEditInfo;
         const fieldWidget = this.allFieldWidgets[this.state.id]
-            .find(field => field[symbol] === id);
+            .find(field => field[symbol].startsWith(fieldName) && field.isFocusable());
```
startsWith works but relies on an implementation detail... When we talked about this earlier I forgot about that. Is it ok to replace `field[symbol].startsWith(fieldName)` by `field.name === fieldName`?

aab-odoo commented on addons/web/static/src/js/fields/relational_fields.js:
```diff
@@ -2835,6 +2835,13 @@ var FieldMany2ManyCheckBoxes = AbstractField.extend({
+     * @returns {jQuery}
+     */
+    getFocusableElement: function () {
+        return this.$el;
```
`this.$el` is enough? Shouldn't focus the first checkbox or something?

mcm-odoo commented:
@odoo/rd-security 
Hello, can you check the security risks?

mart-e commented:
> Hello, can you check the security risks?

@mcm-odoo could you please detail why the 4 warning found by the runbot are safe? Where does the data comes from and are we sure the user input are escaped in all cases?
It will make our review easier with a bonus [rubber duck debugging](https://en.wikipedia.org/wiki/Rubber_duck_debugging) ;-)

mcm-odoo commented:
Hello @mart-e 
These two cases do the same thing
```js
this.$('.o_field_many2one_extra').html(this._renderValueLines(false));
// and
this.$el.html(this._renderValueLines(true));
```
They are in the render functions and use the string returned by `this._renderValueLines` which is implemented like this
```js
_renderValueLines: function (needFirstLine) {
        const escapedValue = _.escape((this.m2o_value || "").trim());
        const lines = escapedValue.split('\n');
        if (!needFirstLine) {
            lines.shift();
        }
        return lines.map((line) => `<span>${line}</span>`).join('<br/>');
},
```
This implementation was at the place of `this.$el.html(this._renderValueLines(true));` before the commit. I just extracted it as is (with an es6 update) and factorized it.

The case `this.$el.html(qweb.render(this.template, {widget: this}));` was already in the function
```js
_renderCheckboxes: function () {
        var self = this;
        this.m2mValues = this.record.specialData[this.name];
        this.$el.html(qweb.render(this.template, {widget: this})); // <-- here
        _.each(this.value.res_ids, function (id) {
            self.$('input[data-record-id="' + id + '"]').prop('checked', true);
        });
}
```
This function was called in the `_renderReadonly` and `_renderEdit` functions but now those 3 render functions have been merged into one `_render` function. I didn't really touch this line because it still the same as before.

And for the case `return $td.html(formattedValue).attr('title', title).attr('name', name);`
Before it was `return $td.html(formattedValue).attr('title', title);`, I just added `.attr('title', title)` to the line.

mart-e commented:
Thanks for the detail information :+1: 

> And for the case `return $td.html(formattedValue).attr('title', title).attr('name', name);`
> Before it was `return $td.html(formattedValue).attr('title', title);`, I just added `.attr('title', title)` to the line.

I agree you haven't introduced it but it's a good opportunity to review it. In this specific case, it is escaped by the `var formattedValue = formatter(value, field, formatOptions);` call.
For `render` the call itself is safe but the template used is the one to be careful of. This is covered by another CI test to catch usages of `t-raw`.

@robodoo override=ci/security
Reason: detailed explanations [above](https://github.com/odoo/odoo/pull/60693#issuecomment-771471104). All the user content is escaped before going to `.html()`

xmo-odoo commented:
> The case `this.$el.html(qweb.render(this.template, {widget: this}));` was already in the function
> 
> [...]
> 
> And for the case `return $td.html(formattedValue).attr('title', title).attr('name', name);`
> Before it was `return $td.html(formattedValue).attr('title', title);`, I just added `.attr('title', title)` to the line.

Note that just because you modify existing code doesn't mean the existing code *is* safe :)

Here it looks fine:

* qweb output should be safe by definition (modulo the template's own unsafeties)
* `formattedValue` is `escape: true`

@ged-odoo though for the latter it might be worth to:
* make escaping opt-out instead of opt-in
* apply an escaper wrapper to every exported formatter unless they're marked specifically
* go over the formatters with a find-toothed comb for the places where we might have historically been too trusting e.g.
  - formatMonetary uses `currency.symbol` as-is and that's normally only updatable by admins but it still seems unnecessarily trusting, assuming currencies are fetched from an external service *that* may get compromised
  - `formatFloat` trusts that `formatterCallback` doesn't fuck up, that seems unnecessary
  - `formatFloat` also assumes `precision` is a number which I am not sure is guaranteed

(I'm virtually certain we have similar issues on the python side, this is not an indictment).

aab-odoo commented:
robodoo rebase-merge r+

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

robodoo commented:
Linked pull request(s) odoo/enterprise#14931 not ready. Linked PRs are not staged until all of them are ready.

randall-vx commented:
@oscarolar
